rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+terminated and the previous installation is replaced.
+
+The first option is an emergent property of the level-triggered system and will be supported from the beginning.
+Teleport won't bother to distinguish between an upgrade and a downgrade. No special downgrade logic is required
+for this option to work.
+
+The second option will require a decent amount of specialized support and will be added later down the line. Script
+installers would likely need to be amended in some way to work correctly with a local rollback scheme. The
+details of how exactly local rollbacks should function are TBD. Some possibilities include:
+
+- Initially install new versions to a pending location (e.g. `/usr/local/bin/teleport.pending`).  Have teleport
+automatically fork a background monitor and `exec`s into the pending binary if it is detected on startup. If the
+background monitor observes that its requirements are met, it moves the pending binary to the active location,
+replacing the previous install.
+
+- Formally embrace the idea of multiple concurrently installed teleport versions
+and provide a thin "proxy binary" that can seamlessly `exec` into the current target version based on some filesystem
+config, potentially launching a background monitor of a different version first depending on said config. This has the
+downside of introducing a new binary, but the upside of eliminating the need for messy move/rename schemes.
+
+- Fully install the new version, creating a backup of the previous version first. Rely on an external mechanism for
+ensuring that the monitor/revert process gets run (e.g. by registering a new systemd unit). This has the upside of
+probably being compatible with script-based installers without any changes (teleport could create the backup and register
+the unit before starting the script), but has the downside of introducing an external dependency.
+
+## Notes
+
+### Open Questions
+
+- How important is establishing a workflow around pending/approval (e.g. have automatic disovery and
+automatic installation, but wait for administrative intervention before actually updating the directive)?
+We want this eventually, but is it of sufficient priority to be included from the beginning?
+
+- How best to support hybrid/mapped `version-controller`s? (e.g. if a user wants to apply some custom
+modification to the output of the TUF controller to better suit their specific needs) Should it be
+possible to emit a pending directive that gets modified and placed in the active slot by a separate
+task?  Should an explicit mapping/modifier component be defined?
+
+- How should client upgrades work? Having clients pull their source of truth from the teleport cluster
+is potentially confusing in multi-cluster environments.  Maybe a TUF-based `tsh upgrade` command?  Do
+we care about supporting alternative/pluggable install for clients?
+
+- How should local-rollbacks work? Presumably some background process will fork and monitor the new
```

Comment:
Teleport has upgrade/rollback sequence - https://goteleport.com/docs/setup/admin/graceful-restarts/#graceful-upgrades we should leverage that